### PR TITLE
Do not crash on duplicated product names

### PR DESCRIPTION
**Pull request by neonichu**
Instead we need to raise an error if we end up in that situation.

rdar://106432785

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 1]
```diff
@@ -435,7 +435,16 @@ private func createResolvedPackages(
...
             })
-        let productDependencyMap = lookupByProductIDs ? productDependencies.spm_createDictionary({ ($0.product.identity, $0) }) : productDependencies.spm_createDictionary({ ($0.product.name, $0) })
+
+        let productDependencyMap: [String: ResolvedProductBuilder]
+        if lookupByProductIDs {
+            productDependencyMap = productDependencies.spm_createDictionary({ ($0.product.identity, $0) })
+        } else {
+            productDependencyMap = try Dictionary(productDependencies.map({ ($0.product.name, $0) }), uniquingKeysWith: { lhs, rhs in
+                let duplicates = productDependencies.filter { $0.product.name == lhs.product.name }
+                throw PackageGraphError.duplicateProduct(product: lhs.product.name, packages: duplicates.map { $0.packageBuilder.package.identity.description })
```
Comment (MaxDesiatov) [inline Sources/PackageGraph/PackageGraph+Loading.swift:445]: formatting nit: this exceeds 120 characters limit set in SwiftFormat config

Comment (neonichu) [reply Sources/PackageGraph/PackageGraph+Loading.swift:445]: 🫠

Comment (MaxDesiatov) [reply Sources/PackageGraph/PackageGraph+Loading.swift:445]: I'm totally fine with keeping it as is, seems like we agreed that reformatting everything at once in a future PR is the way to go 🫣

Comment (neonichu) [reply Sources/PackageGraph/PackageGraph+Loading.swift:445]: Sorry, wasn't a reaction to you but to the arbitrary SwiftFormat config. Thanks for pointing it out!

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by neonichu** [comment]
@swift-ci please smoke test macOS

**Comment by etcwilde** [comment]
Windows failure:
```
fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Found related pull requests: []
```

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Pull request merged**